File: extensions/music/src/main/java/app/morphe/extension/music/patches/ChangeHeaderPatch.java
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to this code.
 */

package app.morphe.extension.music.patches;

import java.util.Objects;

import app.morphe.extension.shared.Logger;
import app.morphe.extension.shared.ResourceType;
import app.morphe.extension.shared.ResourceUtils;
import app.morphe.extension.music.settings.Settings;

@SuppressWarnings("unused")
public class ChangeHeaderPatch {

    public enum HeaderLogo {
        DEFAULT(null),
        MORPHE("morphe_header_dark"),
        CUSTOM("morphe_header_custom_dark");

        private final String drawableName;

        HeaderLogo(String drawableName) {
            this.drawableName = drawableName;
        }

        private Integer getDrawableId() {
            if (drawableName == null) {
                return null;
            }

            int id = ResourceUtils.getIdentifier(ResourceType.DRAWABLE, drawableName);
            if (id == 0) {
                Logger.printException(() ->
                        "Header drawable not found: " + drawableName
                );
                Settings.HEADER_LOGO.resetToDefault();
                return null;
            }

            return id;
        }
    }

    /**
     * Injection point.
     */
    public static int getHeaderDrawableId(int original) {
        return Objects.requireNonNullElse(
                Settings.HEADER_LOGO.get().getDrawableId(),
                original
        );
    }
}

File: patches/src/main/kotlin/app/morphe/patches/music/layout/branding/header/ChangeHeaderPatch.kt
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to this code.
 */

package app.morphe.patches.music.layout.branding.header

import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.extensions.InstructionExtensions.getInstruction
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.shared.layout.branding.header.baseChangeHeaderPatch
import app.morphe.patches.all.misc.resources.ResourceType
import app.morphe.patches.all.misc.resources.getResourceId
import app.morphe.patches.all.misc.resources.resourceMappingPatch
import app.morphe.patches.music.misc.settings.PreferenceScreen
import app.morphe.patches.music.shared.Constants.COMPATIBILITY_YOUTUBE_MUSIC
import app.morphe.util.forEachLiteralValueInstruction
import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction

private val targetResourceDirectoryNames = mapOf(
    "drawable-hdpi" to "121x36 px",
    "drawable-xhdpi" to "160x48 px",
    "drawable-xxhdpi" to "240x72 px",
    "drawable-xxxhdpi" to "320x96 px"
)

private val variants = arrayOf("dark")
private val logoResourceNames = arrayOf("morphe_header_dark")

private val headerDrawableNames = arrayOf(
    "action_bar_logo_ringo2",
    "ytm_logo_ringo2"
)

private const val EXTENSION_CLASS =
    "Lapp/morphe/extension/music/patches/ChangeHeaderPatch;"

private val changeHeaderBytecodePatch = bytecodePatch {
    dependsOn(resourceMappingPatch)

    execute {
        headerDrawableNames.forEach { drawableName ->
            val drawableId = getResourceId(ResourceType.DRAWABLE, drawableName)

            forEachLiteralValueInstruction(drawableId) { literalIndex ->
                val register = getInstruction<OneRegisterInstruction>(literalIndex).registerA

                addInstructions(
                    literalIndex + 1,
                    """
                        invoke-static { v$register }, $EXTENSION_CLASS->getHeaderDrawableId(I)I
                        move-result v$register
                    """
                )
            }
        }
    }
}

@Suppress("unused")
val changeHeaderPatch = baseChangeHeaderPatch(
    targetResourceDirectoryNames = targetResourceDirectoryNames,
    variants = variants,
    logoResourceNames = logoResourceNames,
    appendVariantToLogo = false,
    preferenceScreen = PreferenceScreen.GENERAL,
    block = {
        dependsOn(changeHeaderBytecodePatch)
        compatibleWith(COMPATIBILITY_YOUTUBE_MUSIC)
    }
)